frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/MacheteFileWriter.java

```diff
@@ -0,0 +1,64 @@
+package com.virtuslab.gitmachete.frontend.file;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.nio.file.Path;
+
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.vfs.VfsUtilCore;
+import com.intellij.openapi.vfs.VirtualFileManager;
+import io.vavr.collection.List;
+import io.vavr.control.Option;
+import io.vavr.control.Try;
+import lombok.CustomLog;
+import lombok.val;
+
+import com.virtuslab.branchlayout.api.BranchLayout;
+import com.virtuslab.branchlayout.api.BranchLayoutException;
+import com.virtuslab.branchlayout.api.readwrite.IBranchLayoutWriter;
+import com.virtuslab.branchlayout.impl.readwrite.BranchLayoutFileUtils;
+import com.virtuslab.qual.guieffect.UIThreadUnsafe;
+
+@CustomLog
+public class MacheteFileWriter {
+
+  @UIThreadUnsafe
+  @SuppressWarnings("regexp") // to allow for `synchronized`
+  public synchronized void writeBranchLayout(Path path, BranchLayout branchLayout, boolean backupOldFile,
+      IBranchLayoutWriter branchLayoutWriter)
+      throws BranchLayoutException {
+    LOG.debug(() -> "Writing branch layout to (${path}), branchLayout = ${branchLayout}, backupOldFile = ${backupOldFile}");
+    val macheteVFile = VirtualFileManager.getInstance().findFileByNioPath(path);
+
+    try {
+      if (macheteVFile != null) {
+        if (backupOldFile) {
+          val parentDir = macheteVFile.getParent();
+          assert parentDir != null : "Can't get parent directory of branch layout file";
+
+          val backupFileName = macheteVFile.getName() + "~";
+          val backupVFile = ReadAction.compute(() -> Option.of(parentDir.findChild(backupFileName)));
+          if (backupVFile.isDefined()) {
+            backupVFile.get().delete(/* requestor */ this);
+          }
+
+          Try.of(() -> VfsUtilCore.copyFile(/* requestor */ this, macheteVFile, parentDir, backupFileName))
+              .getOrElseThrow(
+                  e -> new BranchLayoutException(
+                      "Unable to backup branch layout file from ${macheteVFile.toNioPath()} to ${parentDir.toNioPath().resolve(backupFileName)}",
+                      e));
+        }
+        LOG.debug(() -> "Deriving IdentSpec from (${macheteVFile.toNioPath()})");
+        val indentSpec = BranchLayoutFileUtils.deriveIndentSpec(
+            List.ofAll(VfsUtilCore.loadText(macheteVFile).lines()));
+
+        try (val outputStream = new BufferedOutputStream(macheteVFile.getOutputStream(/* requestor */ this))) {
+          branchLayoutWriter.write(outputStream, branchLayout, indentSpec);
+        }
+      }
+
+    } catch (IOException e) {
+      throw new BranchLayoutException("Error while writing to (${macheteVFile.toNioPath()})", e);
```

Comment:
Hmm I wouldn't throw `BranchLayoutException` at this level... in fact, just letting `IOException` be thrown up the stack doesn't look like a bad idea

Reply:
This would also save you a `Try.of` above